### ISSUE TITLE
stack overflow with big lists

### ISSUE THREAD

**Issue by toolslive** [mjambon/yojson#47]
```
let r n =
      let rec loop acc i = 
        if i = 0 then `List acc
        else loop (`Int i :: acc) (i-1)
      in loop [] n;;
# let big = r 1_000_000;;
val big :...
# let r_s = Yojson.Safe.pretty_to_string big;;
Stack overflow during evaluation (looping recursion?).
```
I think it boils down to List.map being non tail recursive.

**Comment by mjambon** [comment]
Thanks for reporting this. Someone would have to look at the source code of yojson and/or easy-format. I don't have much time at the moment. `Yojson.Safe.pretty_to_string` maps the json tree to an easy-format tree, which easy-format then converts to a string.
If it turns out it's too difficult to fix, a workaround may be to provide a faster, simpler json pretty-printer in addition to the current, fancy one. Easy-format uses the Format module and it can be very CPU-intensive. So even if it terminates without crashing, it may still be prohibitively slow.

**Comment by toolslive** [comment]
Well, I looked at it a bit. There's a problem here, and a problem in https://github.com/mjambon/easy-format (Fortunately, you know that author)
  - `List.map` used in pretty.ml isn't tail recursive
  - `List.map` and `List.split` in easy_format.ml aren't tail recursive.

The solution is something like this:
```
let safe_map f xs =
  let rec helper acc = function
    | [] -> List.rev acc
    | x :: xs ->
       helper (f x :: acc) xs
  in
  helper [] xs


let safe_rev_split xys =
  let rec helper xs ys = function
    | [] -> xs, ys
    | (x, y) :: xys -> helper (x :: xs) (y :: ys) xys
  in
  helper [] [] xys

let safe_split xys = safe_rev_split (List.rev xys)
```
and then replace all occurrences of `List.map` with `safe_map`
and all occurrences of `List.split` with `safe_split`

PS.
one cannot but appreciate the irony of this: https://github.com/ocaml/ocaml/pull/867

If you want, I can make pull requests in both libraries with the proposed changes. 
I could also expose the function in `easy-format`'s api so `safe_map` doesn't need to be added twice,
but it really belongs in the standard lib.

**Comment by andersfugmann** [comment]
Opened PR https://github.com/mjambon/easy-format/pull/18 to fix the issue.

**Comment by mjambon** [comment]
Thanks guys for taking care of this. @andersfugmann fixed the problem in easy-format. The fix is in easy-format 1.3.1, which will be in opam shortly, see https://github.com/ocaml/opam-repository/pull/11216.
